### PR TITLE
10.0 access restricted

### PR DESCRIPTION
**Pull request by iledarn**

### CODE REVIEW

File: access_apps/static/src/js/dashboard.js [hunk 1]
```diff
@@ -4,12 +4,21 @@ odoo.define('access_apps.dashboard', function (require) {
     var dashboard = require('web_settings_dashboard');
 
     dashboard.Dashboard.include({
-        load_apps: function(data){
-            if(data.has_access_to_apps) {
-                return this._super(data);
-            } else {
-                this.$('.o_web_settings_dashboard_apps').parent().remove();
-            }
```
Comment (yelizariev) [inline access_apps/static/src/js/dashboard.js]: I don't see problem with previous solution. No need to refactor. Using session_info is too intrusive

**Comment by yelizariev** [comment]
```
odoo.define('access_apps.dashboard', function (require) {
"use strict";

var def = $.Deffered()
    var dashboard = require('web_settings_dashboard');

    dashboard.Dashboard.include({
        load_apps: function(data){
            if(data.has_access_to_apps) {
                return this._super(data).then(function(){
def.resolve();
});
            }
            this.$('.o_web_settings_dashboard_apps').parent().remove();
def.resolve();

        }
   });
 return {'ready': def}
});

```

File: access_apps/static/src/js/dashboard.js [hunk 2]
```diff
@@ -2,14 +2,25 @@ odoo.define('access_apps.dashboard', function (require) {
...
+
+        willStart:function(){
+            var self = this;
+	    var Users = new Model('res.users');
+		return Users.call('has_group', ['access_apps.group_allow_apps']).then(function (result) {
```
Comment (yelizariev) [inline access_apps/static/src/js/dashboard.js]: Such solution delays dashboard showing

File: access_apps/tests/test_js.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+@odoo.tests.common.post_install(True)
+class TestUi(odoo.tests.HttpCase):
+
+    def test_01_dashboard_remove(self):
+        demo_user = self.env.ref('base.user_demo')
+        system_group = self.env.ref('base.group_system')
+        allow_apps_group = self.env.ref('access_apps.group_allow_apps')
+        demo_user.write({'groups_id': [(4, system_group.id)]})
+        demo_user.write({'groups_id': [(3, allow_apps_group.id)]})
+        self.phantom_js("/", "odoo.__DEBUG__.services['web_tour.tour'].run('removed_apps_dashboard')", "odoo.__DEBUG__.services['access_apps.dashboard'].ready.state() == 'resolved'", login="demo")
```
Comment (yelizariev) [inline access_apps/tests/test_js.py]: https://odoo-development.readthedocs.io/en/latest/dev/tests/js.html#phantom-js-python-tests

File: access_apps/static/src/js/access_apps_tour_dashboard.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    var tour = require("web_tour.tour");
+
+    tour.register('removed_apps_dashboard', {
+        test: true,
+        url: '/web',
+    },
+        [
+            {
+                trigger: '.o_app[data-menu="4"], .oe_menu_toggler[data-menu="4"]',
```
Comment (yelizariev) [inline access_apps/static/src/js/access_apps_tour_dashboard.js]: Use ``data-menu-xmlid`` instead

**Comment by yelizariev** [comment]
[ ] add this to travis.yml:

     - PYLINT_ODOO_JSLINTRC="/home/travis/maintainer-quality-tools/travis/cfg/.jslintrc"

https://github.com/it-projects-llc/misc-addons/blob/11.0/.travis.yml#L19

**Comment by ursus102** [comment]
Under the account 'Administrator' I selected an empty line in the field 'Apps access' and save. After this action, there is no way to use Odo required rebuild

> Odoo Session Expired
Traceback (most recent call last):
  File "/mnt/odoo-extra/runbot/static/build/10800-66-4bce29/odoo/http.py", line 640, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/mnt/odoo-extra/runbot/static/build/10800-66-4bce29/odoo/addons/base/ir/ir_http.py", line 184, in _dispatch
    auth_method = cls._authenticate(func.routing["auth"])
  File "/mnt/odoo-extra/runbot/static/build/10800-66-4bce29/odoo/addons/base/ir/ir_http.py", line 112, in _authenticate
    getattr(cls, "_auth_method_%s" % auth_method)()
  File "/mnt/odoo-extra/runbot/static/build/10800-66-4bce29/odoo/addons/base/ir/ir_http.py", line 85, in _auth_method_user
    raise http.SessionExpiredException("Session expired")
SessionExpiredException: Session expired

**Comment by yelizariev** [comment]
[ ] Superuser must always have access to Settings

**Comment by yelizariev** [comment]
@ursus102 I cannot reproduce the issue. Are you sure, that `` Session expired`` is reproduced after steps you specified?

**Comment by ursus102** [comment]
@yelizariev Now I could not reproduce. But this time I installed "Apps access" in "Allow install apps only from settings" and again got an error. Full chronology:
1. Added a new user - just test
2. I went to account "Administrator"
3. Changed the value of "Apps access" in "Allow install apps only from settings"
4. Error (default)
5. And even the cancellation of actions does not eliminate the error.

**Comment by yelizariev** [comment]
@ursus102 I can only reproduce it if I logout at another tab before clicking Save

File: access_restricted/doc/index.rst [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+Let's take ``Sale`` module as an example.
+
+Without this module installed:
+
+* Say you have a user with administration rights ``Administration: Access Rights``. This user thus may increase his own rights in ``Application > Sales`` from ``Sales: User: All Documents``
+  to ``Sales: Manager``. Also he can open menu ``[[ Sales ]] >> Configuration >> Settings`` and select ``Display 3 fields on sales orders: customer, invoice address, delivery address`` there
+  and then click ``[Apply]`` button (adding ``group_sale_delivery_address``)
+
+With this module installed:
```
Comment (ursus102) [inline access_restricted/doc/index.rst]: 1. Installed the module Sales
2. Installed the user Demo checkbox "Allow add implied groups from settings"
3. User: All Documents
4. Administration: Access Rights
5. I go under the user and do not see Sales - Configuration - Settings.

Administration: Access Rights is not enough to show the Settings in Sales - you need to set the Administration: Settings.
In this case, if the "Allow add implied groups from settings" checkbox is not active then in the Sales - Configuration - Settings you can still see the radio address "Addresses:
- Invoicing and shipping addresses are always the same (Example: services companies)
- Display 3 fields on sales orders: customer, invoice address, delivery address "

Comment (ursus102) [inline access_restricted/doc/index.rst]: 1. Installed the module Sales
2. Installed the user Demo checkbox "Allow add implied groups from settings"
3. User: All Documents
4. Administration: Access Rights
5. I go under the user and do not see Sales - Configuration - Settings.

Administration: Access Rights is not enough to show the Settings in Sales - you need to set the Administration: Settings.
In this case, if the "Allow add implied groups from settings" checkbox is not active then in the Sales - Configuration - Settings you can still see the radio address "Addresses:
- Invoicing and shipping addresses are always the same (Example: services companies)
- Display 3 fields on sales orders: customer, invoice address, delivery address "

**Comment by yelizariev** [comment]
We probably need to imply ``Adminstration / Settings`` rights when user has access ``Allow add implied groups from settings``, otherwise latter doesn't have sense, because user doens't have access to Settings. see https://github.com/it-projects-llc/access-addons/pull/66#discussion_r151070397

**Comment by yelizariev** [comment]
For squashed commit:

[ADD] set of updates
```

access_apps:

`1.2.0`
-------

- REF: rename "Show Apps Menu" to "Allow install apps"
- ADD: "Allow install apps only from settings"
- IMP: group "Show Apps Menu" and "Allow install apps only from settings" under "Apps access" security category
 

access_restricted:

`1.3.0`
-------

- [ADD] security group that allows increasing rights from settings menu (by checking ``res.config.settings`` 'group_XXX' boolean fields)

```

**Comment by yelizariev** [comment]
- [ ] write tests for  access_restricted (User has access -> no errors, User has not access -> error is raised)

**Comment by yelizariev** [comment]
- [ ] coverage for methods in ``'res.users`` and ``res.groups`` 
  * emulate opening of res.user form -- repeat call that browser does via xmlrpc when you open view manulally
  * add new groups to users and open form again -- new group must be appeared at the form

**Comment by iledarn** [comment]
@yelizariev please check 329739d for > test that non-super admin don't see unavailable fields (groups)

**Comment by iledarn** [comment]
@yelizariev 
see #66 (comment)

**Comment by iledarn** [comment]
@yelizariev ,
> write tests for new features of access_restricted (User has access -> no errors, User has not access -> error is raised)

This is also done, please merge

File: access_restricted/tests/test_allow_implied.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        group_allow = self.env.ref('access_restricted.group_allow_add_implied_from_settings')
+        demo_user.write({'groups_id': [(4, group_allow.id)]})
+
+        # check that now the field is not readonly
+        self.assertFalse(test_config_settings.fields_get()['group_user']['readonly'])
+        # check that now the group is in classified
+        self.assertTrue(test_config_settings._get_classified_fields()['group'])
+
+        test_config_settings.sudo(demo_user.id).execute()
+        self.assertTrue(self.env['res.users'].sudo(demo_user.id).has_group('base.group_user'))
```
Comment (yelizariev) [inline access_restricted/tests/test_allow_implied.py]: Add check that demo user doesn't have group_user before calling ``execute()``

**Comment by iledarn** [comment]
@yelizariev

**Pull request merged**